redis_session.c

```diff
@@ -183,6 +185,162 @@ redis_pool_get_sock(redis_pool *pool, const char *key TSRMLS_DC) {
     return NULL;
 }
 
+int lock_acquire(RedisSock *redis_sock, redis_session_lock_status *lock_status TSRMLS_DC)
+{
+    if (lock_status->is_locked || !INI_INT("redis.session.locking_enabled")) return SUCCESS;
+
+    char *cmd, *response;
+    int response_len, cmd_len, lock_wait_time, max_lock_retries, i_lock_retry, lock_expire;
+    calculate_lock_secret(lock_status);
+
+    lock_wait_time = INI_INT("redis.session.lock_wait_time");
+    if (lock_wait_time == 0) {
+      lock_wait_time = 2000;
+    }
+
+    max_lock_retries = INI_INT("redis.session.lock_retries");
+    if (max_lock_retries == 0) {
+      max_lock_retries = 10;
+    }
+
+    lock_expire = INI_INT("redis.session.lock_expire");
+    if (lock_expire == 0) {
+      lock_expire = INI_INT("max_execution_time");
+    }
+
+    // Building the redis lock key
+    smart_string_appendl(&lock_status->lock_key, lock_status->session_key, strlen(lock_status->session_key));
+    smart_string_appendl(&lock_status->lock_key, "_LOCK", strlen("_LOCK"));
+    smart_string_0(&lock_status->lock_key);
+
+    if (lock_expire > 0) {
+        cmd_len = REDIS_SPPRINTF(&cmd, "SET", "ssssd", lock_status->lock_key.c, lock_status->lock_key.len, lock_status->lock_secret.c, lock_status->lock_secret.len, "NX", 2, "PX", 2, lock_expire * 1000);
+    } else {
+        cmd_len = REDIS_SPPRINTF(&cmd, "SET", "sss", lock_status->lock_key.c, lock_status->lock_key.len, lock_status->lock_secret.c, lock_status->lock_secret.len, "NX", 2);
+    }
+
+    for (i_lock_retry = 0; !lock_status->is_locked && (max_lock_retries == -1 || i_lock_retry <= max_lock_retries); i_lock_retry++) {
+      if(!(redis_sock_write(redis_sock, cmd, cmd_len TSRMLS_CC) < 0)
+          && ((response = redis_sock_read(redis_sock, &response_len TSRMLS_CC)) != NULL)
+          && response_len == 3
+          && strncmp(response, "+OK", 3) == 0) {
+            lock_status->is_locked = 1;
+      } else if (max_lock_retries == -1 || i_lock_retry < max_lock_retries) {
+          usleep(lock_wait_time);
+      }
+    }
+
+    if (response != NULL) {
+        efree(response);
+    }
+    efree(cmd);
+
+    if (lock_status->is_locked) {
+        return SUCCESS;
+    } else {
+        return FAILURE;
+    }
+}
+
+void refresh_lock_status(RedisSock *redis_sock, redis_session_lock_status *lock_status TSRMLS_DC)
+{
+    if (!lock_status->is_locked) return;
+    // If redis.session.lock_expire is not set => TTL=max_execution_time
+    // Therefore it is guaranteed that the current process is still holding the lock
+    if (lock_status->is_locked && INI_INT("redis.session.lock_expire") == 0) return;
+
+    char *cmd, *response;
+    int response_len, cmd_len;
+
+    cmd_len = REDIS_SPPRINTF(&cmd, "GET", "s", lock_status->lock_key.c, lock_status->lock_key.len);
+
+    redis_sock_write(redis_sock, cmd, cmd_len TSRMLS_CC);
```

Comment:
Don't ignore return value of redis_sock_write

Reply:
Will be fixed. I think in addition a warning makes sense.
In case of error (response == null), is_locked is set to false anyway and write operation will be denied.